### PR TITLE
Split NR and outer loops max iterations

### PR DESCRIPTION
**Pull request by geofjamg**
Signed-off-by: Geoffroy Jamgotchian <geoffroy.jamgotchian@gmail.com>

**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
No

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
Feature

**What is the current behavior?** *(You can also link to an open issue here)*
We can only control the max number of cumulated NR iterations.

**What is the new behavior (if this is a feature change)?**
We can control the max number of iterations for each NR run and the max number of outer loop iterations. This will allow to better adapt the max number of outer loop iterations when more and more controls (outer loops) are added without changing the NR configuration.

**Does this PR introduce a breaking change or deprecate an API?** *If yes, check the following:*
- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration guide has been updated in the github wiki *(What changes might users need to make in their application due to this PR?)*

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

File: src/main/java/com/powsybl/openloadflow/ac/AcloadFlowEngine.java [hunk 1]
```diff
@@ -146,12 +158,9 @@ public AcLoadFlowResult run() {
             outerLoop.cleanup(outerLoopContext);
         }
 
-        int nrIterations = runningContext.lastNrResult.getIteration();
-        int outerLoopIterations = runningContext.outerLoopIterationByType.values().stream().mapToInt(MutableInt::getValue).sum() + 1;
-
         AcLoadFlowResult result = new AcLoadFlowResult(context.getNetwork(),
```
Comment (geofjamg) [inline src/main/java/com/powsybl/openloadflow/ac/AcloadFlowEngine.java:161]: To fix: add a status for max outer loop iterations reached.

Comment (geofjamg) [reply src/main/java/com/powsybl/openloadflow/ac/AcloadFlowEngine.java:164]: Done

File: src/main/java/com/powsybl/openloadflow/ac/AcLoadFlowResult.java [hunk 2]
```diff
@@ -11,13 +11,15 @@
...
+import java.util.Objects;
+
 /**
  * @author Geoffroy Jamgotchian <geoffroy.jamgotchian at rte-france.com>
  */
 public class AcLoadFlowResult extends AbstractLoadFlowResult {
 
     public static AcLoadFlowResult createNoCalculationResult(LfNetwork network) {
-        return new AcLoadFlowResult(network, 0, 0, NewtonRaphsonStatus.NO_CALCULATION, Double.NaN, Double.NaN);
+        return new AcLoadFlowResult(network, 0, 0, NewtonRaphsonStatus.NO_CALCULATION, OuterLoopStatus.STABLE, Double.NaN, Double.NaN);
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/ac/AcLoadFlowResult.java:22]: New enum for outer loop status?

File: src/main/java/com/powsybl/openloadflow/ac/nr/NewtonRaphson.java [hunk 3]
```diff
@@ -64,8 +63,8 @@ public static List<Pair<Equation<AcVariableType, AcEquationType>, Double>> findL
                 .collect(Collectors.toList());
     }
 
-    private NewtonRaphsonStatus runIteration(StateVectorScaling svScaling) {
-        LOGGER.debug("Start iteration {}", iteration);
+    private NewtonRaphsonStatus runIteration(StateVectorScaling svScaling, MutableInt iterations) {
+        LOGGER.debug("Start iteration {}/{}", iterations, parameters.getMaxIterations());
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/ac/nr/NewtonRaphson.java:67]: This is weird: the max iterations is not always the total number of iterations.

File: src/main/java/com/powsybl/openloadflow/ac/AcLoadFlowResult.java (hunk 2 above)
Comment (geofjamg) [reply src/main/java/com/powsybl/openloadflow/ac/AcLoadFlowResult.java:22]: I'm not sure it worth it

File: src/main/java/com/powsybl/openloadflow/ac/nr/NewtonRaphsonParameters.java [hunk 4]
```diff
@@ -13,23 +13,19 @@
  */
 public class NewtonRaphsonParameters {
 
-    public static final int DEFAULT_MAX_ITERATION = 30;
+    public static final int DEFAULT_MAX_ITERATIONS = 15;
```
[the remark this answers is not in the document]
Comment (geofjamg) [reply src/main/java/com/powsybl/openloadflow/ac/nr/NewtonRaphsonParameters.java:16]: 10 might be too low for very rare and difficult cases, 20 is probably too much so 15...

File: src/main/java/com/powsybl/openloadflow/ac/nr/NewtonRaphson.java (hunk 3 above)
Comment (geofjamg) [reply src/main/java/com/powsybl/openloadflow/ac/nr/NewtonRaphson.java]: Removed

File: src/main/java/com/powsybl/openloadflow/ac/AcloadFlowEngine.java [hunk 5]
```diff
@@ -54,6 +54,10 @@ private static class RunningContext {
         private NewtonRaphsonResult lastNrResult;
 
         private final Map<String, MutableInt> outerLoopIterationByType = new HashMap<>();
+
+        private int cumulatedOuterLoopIterations = 0;
```
[the remark this answers is not in the document]
Comment (geofjamg) [reply src/main/java/com/powsybl/openloadflow/ac/AcloadFlowEngine.java:58]: after discussions -> total

File: src/main/java/com/powsybl/openloadflow/ac/AcloadFlowEngine.java [hunk 6]
```diff
@@ -118,25 +124,30 @@ public AcLoadFlowResult run() {
...
         double initialSlackBusActivePowerMismatch = runningContext.lastNrResult.getSlackBusActivePowerMismatch();
 
+        runningContext.nrTotalIterations.add(runningContext.lastNrResult.getIterations());
+
         // continue with outer loops only if initial Newton-Raphson succeed
         if (runningContext.lastNrResult.getStatus() == NewtonRaphsonStatus.CONVERGED) {
 
             // re-run all outer loops until Newton-Raphson failed or no more Newton-Raphson iterations are needed
-            int oldIterationCount;
+            int oldCumulatedNrIterations;
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/ac/AcloadFlowEngine.java:133]: You have kept here oldCumulatedNrIterations.

Comment (geofjamg) [reply src/main/java/com/powsybl/openloadflow/ac/AcloadFlowEngine.java]: Fixed

File: src/main/java/com/powsybl/openloadflow/network/impl/Transformers.java [hunk 7]
```diff
@@ -187,13 +188,13 @@ public static int findTapPosition(PhaseTapChanger ptc, double angle) {
...
      */
-    public static int findTapPosition(RatioTapChanger rtc, double ptcRho, double rho) {
+    public static OptionalInt findTapPosition(RatioTapChanger rtc, double ptcRho, double rho) {
         for (int position = rtc.getLowTapPosition(); position <= rtc.getHighTapPosition(); position++) {
             if (Math.abs(rho - ptcRho * rtc.getStep(position).getRho()) < EPS_ALPHA) {
-                return position;
+                return OptionalInt.of(position);
             }
         }
-        throw new PowsyblException("No tap position found (should never happen)");
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/network/impl/Transformers.java:196]: Do we need to log an error message or not? It is the issue I met last week but I am not sure to understand why we can have no tap position...

**Review by annetill — APPROVED** [review]
Maybe just a log but all is okay!

**Pull request merged**